crossplane/crossplane
# Support running Containerized Composition functions

### Description of your changes


Fixes https://github.com/crossplane/crossplane/issues/2524
Closes https://github.com/crossplane/crossplane/pull/3550

This PR implements Composition Functions per the [design document](https://github.com/crossplane/crossplane/blob/b9e4773dc431d974a44232b95500e6f815a9b8ac/design/design-doc-composition-functions.md).

It introduces a new OCI image - `crossplane/xfn` - with a corresponding `xfn` binary.

* `xfn start` starts a long-running process, listening on a gRPC API for requests to run . 
* `xfn run` is a test/helper utility that runs a single Composition Function.

I made a simple no-op Composition Function that can be used to test `xfn`:

```console
$ cat ~/fnio.yaml | xfn run negz/xfn-nop:latest -
```

In order to run this Composition Function, `xfn` will create a new user (and mount) namespace. It will then invoke `xfn spark` as a user that is root inside the new user namespace, but unprivileged relative to the 'real' parent namespace. Spark will:

1. Fetch and cache the Composition Function image.
2. Build an OCI runtime bundle for the Composition Function container run.
3. Use an OCI runtime (by default `crun`) to run the Composition Function container.
4. Clean everything up.

If `xfn` is run with CAP_SETUID and CAP_SETGID it's able to create a slightly "better" user namespace, in that the user namespace supports up to 65,536 UIDs and GIDs. It's possible to run `xfn` without any privileges at all,  but it will be limited to creating user namespaces where only one UID and GID (the user namespace root's) exist.

If `xfn spark` is running in an environment that supports overlayfs in user namespaces (i.e. Kernel 5.11 or above, with a cache directory that is not itself on an overlayfs) it will use overlayfs to create container root filesystems. Otherwise, it will fall back to an "uncompressed" image cache that extracts uncompressed tarballs each time a function is run. The overlay is faster, while the uncompressed approach is widely compatible.

This PR also implements the "Crossplane end" of Composition Functions - new Composition logic that is capable of running a pipeline of Composition Functions as well as contemporary Patch & Transform (P&T) style Composition. While this change required significant refactoring of the Composition logic I've tried to keep things _mostly_ the same unless `--enable-composition-functions` is set.

One notable place this implementation differs from the design document is that Composition Functions must include _all_ desired state. The design document had proposed they could omit desired state they were not concerned with. I couldn't find a good way to make that design work in practice - in particular it wasn't possible to differentiate "delete this resource (or field)"  from "I don't have an opinion about this resource (or field)".

Unless anything comes up in review, the only remaining blocker to ship this with v1.11 is writing documentation.

There's a few small things I may attempt to add, In separate PRs, before v1.11 if there is time:

- [ ] Implement image pull auth
- [ ]  Have RunFunction return errors with meaningful gRPC status codes
- [ ]  Refactor spark.Run for reduced complexity, add tests
- [ ]  Figure out broken FunctionIO CRD generation (maybe due to missing object metadata?)

Post v1.11 I would like to:

- [ ] Implement some kind of garbage collection for the image cache
- [ ]  Look into optionally deploying xfn as a deployment - it should scale horizontally pretty well.
- [ ]  Determine whether we need to do anything to support multi-platform OCI images.
- [ ]  Allow functions to access the environment (perhaps once its non-alpha?)
- [ ]  Optionally cache results of function runs with the same input for configurable duration

I have:


### How has this code been tested



[contribution process]: https://git.io/fj2m9

This PR needs much more testing, but as of right now Composition Functions are running end-to-end in kind for me. I've also validated that everything appears to work as normal when Composition Functions are not enabled by deploying https://marketplace.upbound.io/configurations/upbound/platform-ref-gcp/v0.3.0.

To test this you can:

1. Checkout this PR.
2. Run `make` to build it
3. Run `./cluster/local/kind.sh up` to get a kind cluster

You can then use the below command to deploy Crossplane with alpha Composition Functions support enabled.

```
HELM3_FLAGS="--set args={--debug,--enable-composition-functions} --set xfn.enabled=true --set xfn.args={--debug}" ./cluster/local/kind.sh helm-install upbound-system
```

Once Crossplane is running you can add a `functions` array to a Composition. I've been using a locally modified fork of platform-ref-gcp to do so.

```yaml
apiVersion: apiextensions.crossplane.io/v1
kind: Composition
metadata:
  name: xgke.gcp.platformref.upbound.io
  labels:
    provider: GCP
spec:
  writeConnectionSecretsToNamespace: upbound-system
  compositeTypeRef:
    apiVersion: gcp.platformref.upbound.io/v1alpha1
    kind: XGKE
  functions:
  - name: nop
    type: Container
    container:
      # This is a function that just returns the desired state that it is passed unmodified
      image: negz/xfn-nop:latest
```

If you want to try write your own function you can see the I/O specification in the [Composition Functions design doc](https://github.com/crossplane/crossplane/blob/9ee7a27dcf643a92df68d2a6de6e55ce95e7f907/design/design-doc-composition-functions.md#function-api). Note as above, and contrary to the design, that the `desired` object of a `FunctionIO` must include _all_ desired state. Your function will be passed a `desired` object containing the desired state according to Crossplane (e.g. from the `resources` array of the Composition if any) and any functions that ran before it. Your function may modify the desired state in any way it sees fit.

```console
$ cat ~/fnio.yaml | docker run \
    -i --rm -u nobody \
    --security-opt seccomp=unconfined \
    --mount type=tmpfs,destination=/xfn \
    build-c4cbe24f/xfn-arm64 run negz/xfn-nop:latest -
```

I've also validated that the above command runs successfully. Note that we don't really have to completely disable seccomp; we just need to allow some extra syscalls such as `unshare` that are not allowed by Docker's (and containerd's) default seccomp policy.

If folks want to help test things out, here are a few ideas:

- [ ] Make sure everything continues to work the same without the `--enable-composition-functions` flag.
- [x] Check that functions can't access the network with `networkPolicy: Isolated`.
- [x] Check that functions can access the network (if the runner can) with `networkPolicy: Runner`.
- [x] A pipeline of multiple functions.
- [ ] Many XRs using a pipeline of multiple functions.
- [x] Writing a function.
